---
title: "proj_test"
author: ""
date: '202X-XX-XX'
output: html_document
---

```{r setup, include=FALSE}
library(reticulate)
library(MetaBAnalysis)
knitr::opts_chunk$set(echo = TRUE)
knitr::knit_engines$set(python = reticulate::eng_python)
```

### Filter the data and run dada2:
```{r}
# Collect sample paths and names. Duplicate line once per primer:
primer1 <- MetaBAnalysis::CollectData("../Filtered_data")

# Filter reads and generate out object for later use:
out <- dada2::filterAndTrim(primer1$Forward, primer1$FiltFs, 
                            primer1$Reverse, primer1$FiltRs,
                            maxN=0, truncQ=2, rm.phix=TRUE,
                            compress=TRUE, multithread=TRUE)
saveRDS(out, file = "out.rds")

# Dereplicate and merge pairs with dada2:
dada2Counts <- MetaBAnalysis::DadaAnalysis(primer1$FiltFs, primer1$FiltRs, 
                                           justConcatenate = FALSE, minOverlap = 5)
```

### Convert and modify  matrix "dada2Counts" to DGEList "yAll":
```{r}
yAll <- MetaBAnalysis::MakeDGEList(dada2Counts, primer1$Samples, primer1$ForwardC)
saveRDS(yAll, file = "yAll.rds")
```

### Export fasta-file:
```{r}
ExportFasta(yAll, "y_test.fa", minLength = 50, maxLength = 1000)
```

### BLAST-tools (Experimental):
```{r}
# blastOut <- MultiBlaster(fastaFile = "y_test.fa")
# BlastResWriter(blastOut, "y_test.out")
```

### Parse BLAST-results:
```{r}
### The following code assumes that the folder "../../Suppl_data" exists and contains the files "compact_names.csv" and
# "compact_nodes.csv. If it does not, create the folder in that location and perform the following:
# 1. $ cd Suppl_data
# 2. $ wget https://ftp.ncbi.nih.gov/pub/taxonomy/taxdmp.zip
# 3. $ unzip taxdmp.zip
# 4. $ awk -F\t '{gsub(/[^[:alnum:]]/," ", $3); if ($7 == "scientific name") {print $1 "\t" $3}}' names.dmp > compact_names.csv
# 5. $ awk -F\t '{gsub(/[^[:alnum:]]/," ", $3); print $1 "\t" $3 "\t" $5}' nodes.dmp > compact_nodes.csv
# 6. $ rm *.dmp *.prt *.txt *.zip

data(compactNameDump, package = "MetaBAnalysis")
data(compactNodeDump, package = "MetaBAnalysis")

blastResY <- MetaBAnalysis::BlastParse(yAll, "y_test.out")
saveRDS(blastResY, file = "blastResY.rds")
```

### Create Summaryfiles from the BLAST-results:
```{r}
allResults <- MetaBAnalysis::SumRes(blastRes = blastResY, counts = yAll$counts, taxGroup = "Fish") # Summarize a list of result objects
```

# Rstudio functions:
```{r}
# These functions are intended to aid the curation step by allowing the user to
# quickly write the sequences related to a species name to a file (QSeqGetter
# and QSSplit) or blast the top three sequences related to a species and View the
# results in Rstudio (QSBLAST).
library(MetaBAnalysis)

# Write sequences to file
QSeqGetter <- function(speciesName, blastResults = blastResY) {
  species_seq <- paste0(">",
                        blastResults[grepl(speciesName, blastResults$species),1],
                        "\n",
                        blastResults[grepl(speciesName, blastResults$species),2])
  writeLines(species_seq, "current_sequence.txt")
  file.edit("current_sequence.txt")
}

# Write sequences to file and split sequences that contains 'NNNNNNNNNN'
QSSplit <- function(speciesName, blastResults = blastResY) {
  seqName <- blastResults[grepl(speciesName, blastResults$species),1]
  seqFull <- blastResults[grepl(speciesName, blastResults$species),2]
  seq1 <- lapply(strsplit(seqFull, "NNNNNNNNNN"), "[[", 1)
  seq2 <- lapply(strsplit(seqFull, "NNNNNNNNNN"), "[[", 2)
  species_seq <- paste0(">", seqName,
                        "\n", seqFull, "\n",
                        ">", seqName, ".1",
                        "\n", seq1, "\n",
                        ">", seqName, ".2",
                        "\n", seq2)
  writeLines(species_seq, "current_sequence.txt")
  file.edit("current_sequence.txt")
}

# Blast the top three sequences of a species
QSBLAST <- function(searchName, seqNumber = 3, blastResults = blastResY) {
  QSeqGetter(searchName, blastResults)
  blastOut <- MultiBlaster("current_sequence.txt",
                           seqNumber,
                           resultNumber = 10,
                           viewChoice = "YES")
  View(blastOut)
}
```

### Functions:
```{r}
### To be added to package:
MakeNotes <- function(countsObject) {
  colnames(countsObject)[colnames(countsObject) == "Species"] <- "Latin"
  outObject <- cbind(Swedish = MetaBAnalysis::TranslateTaxa(countsObject$Latin),
                     countsObject)
  outObject <- cbind(outObject[, c(1, 2)],
                     Percent_all = SpeciesPercent(countsObject),
                     outObject[, -c(1, 2)])
  outObject <- cbind(Notes = rep("", nrow(outObject)), outObject)
  outObject <- cbind(Reference = paste0("hit", 1:nrow(outObject)), outObject)
  return(outObject)
}
  
AddNote <- function(note, reference, countsObject) {
  countsObject$Notes[countsObject$Reference == reference] <- note
  return(countsObject)
}

TrimCutoffs <- function(countsObject, total = 0, within = 0) {
  outObject <- countsObject[countsObject$Percent_all >= total,]
  outObject <- RemoveLowFreqSeqs(outObject, within)
  outObject <- outObject[rowSums(outObject[, -c(1:5)]) > 0,] # remove empty rows
  return(outObject)
}

### To be updated in package:
RenameSpecies <- function (oldName, newName, dataFrame) {
  dataFrame$Latin[grepl(oldName, dataFrame$Latin)] <- newName # one changed row
  return(dataFrame)
}

CombineSpecies <- function(ref1, ref2, dataFrame) {
  speciesSum <- dataFrame[ref1 == dataFrame$Reference, 5:ncol(dataFrame)] +
    dataFrame[ref2 == dataFrame$Reference, 5:ncol(dataFrame)]
  speciesSum <- cbind(dataFrame[ref1 == dataFrame$Reference, 1:4], speciesSum)
  names(speciesSum)[1:4] <- names(dataFrame[1:4])
  dataFrame <- dataFrame[ref1 != dataFrame$Reference,]
  dataFrame <- dataFrame[ref2 != dataFrame$Reference,]
  dataFrame <- rbind(dataFrame, speciesSum)
  dataFrame <- dataFrame[order(rowSums(dataFrame[, -c(1:5)]), decreasing = TRUE),]
  return(dataFrame)
}

RemoveLowFreqSeqs <- function(countsObject, threshold) {
  threshold = threshold / 100 # convert percentage into proportion
  DoCutoff <- function(column, cutoff) {
    column[prop.table(column) < cutoff] <- 0
    return(column)
  }
  counts <- apply(countsObject[, -c(1:5)], MARGIN = 2, DoCutoff, threshold)
  return(cbind(countsObject[, c(1:5)], counts))
}
```

### Curate the results:
```{r}
library(MetaBAnalysis)
library(rvest)

{
  ### Create object:
  groupNotes <- MakeNotes(allResults$TargetGroup)
  
  ### Take notes:
  groupNotes <- AddNote("note", "hitX", groupNotes)
  
  ### Curate:
  # groupNotes <- RenameSpecies("old latin name", "new latin name", groupNotes)
  # groupNotes <- CombineSpecies("hit to be kept", "hit to be absorbed", groupNotes)
  groupNotes[grepl("hitX", groupNotes$Reference), 1] <- "new common name"
  
  ### Filter:
  # groupNotes <- groupNotes[!grepl("drop", groupNotes$Notes),] # Drop all species that has the word "drop" in their notes
  # groupNotes <- groupNotes[rowSums(groupNotes[, -c(1:5)]) > 100,] # remove species with a sequence sum less than 100
  # groupNotes <- TrimCutoffs(total = 0.5, within = 0.5, groupNotes) # trim species based on percentage cutoffs
  
  ### Create outputs:
  groupClean <- groupNotes[, -c(1, 2, 3)]
  columnNames <- colnames(groupClean)
  # columnNames[grepl("old sample name", columnNames)] <- "new sample name"
  names(groupClean) <- columnNames
  allResults$Clean <- groupClean
}

saveRDS(allResults, "allResults.rds") # save the clean object to be used in the report
```

### Generate information about the sequences for use in report:
```{r}
library(rvest)

rawSeqsTable <- MetaBAnalysis::getRawSeqs()

rawSeqs <- rawSeqsTable$nreads

sum(rawSeqs)
min(rawSeqs)
max(rawSeqs)

primerSeqs <- sum(out[, 1])
primerSeqs / sum(rawSeqs) # ratio of sequences left after primer trimming and filtration

groupSeqs <- sum(groupClean[, -c(1, 2)])
groupSeqs / primerSeqs # ratio of filtered sequences belonging to target group

sampleMetaData <- list(rawSeqs = rawSeqs, primerSeqs = primerSeqs, groupSeqs = groupSeqs)
saveRDS(sampleMetaData, "sampleMetaData.rds")

sum(groupClean[1:4, -c(1, 2)]) / sum(groupClean[, -c(1, 2)]) # see what percentage of counts the top four species accounts for
```

### Generate excel-sheet with counts from "groupClean":
```{r}
library(writexl)
write_xlsx(groupClean, "projekt_xyz_sekvensantal.xlsx")
```
