---
title: "proj_test"
author: ""
date: '202X-XX-XX'
output: html_document
---

```{r setup, include=FALSE}
library(reticulate)
knitr::opts_chunk$set(echo = TRUE)
knitr::knit_engines$set(python = reticulate::eng_python)
```

### Filter the data and run dada2:
```{r}
library(dada2)
library(edgeR)

DadaAnalysis <- function(forward, reverse) {
  errF <- dada2::learnErrors(forward, multithread = TRUE)
  errR <- dada2::learnErrors(reverse, multithread = TRUE)
  derepsF <- dada2::derepFastq(forward)
  derepsR <- dada2::derepFastq(reverse)
  dadaF <- dada2::dada(derepsF, err = errF, multithread = TRUE)
  dadaR <- dada2::dada(derepsR, err = errR, multithread = TRUE)
  mergers <- dada2::mergePairs(dadaF, derepsF, dadaR, derepsR, verbose = TRUE)
  seqTab <- dada2::makeSequenceTable(mergers)
  seqTabNoChim <- dada2::removeBimeraDenovo(seqTab, method = "consensus", multithread = TRUE, verbose = TRUE)
  return(seqTabNoChim)
}

forward <- list.files("../Filtered_data", pattern = "_1.fastq.gz", full.names = TRUE) # Gets all the names ending with _1 with "full" path
reverse <- list.files("../Filtered_data", pattern = "_2.fastq.gz", full.names = TRUE) # Gets -"- ending with _2
forwardC <- list.files("../Filtered_data", pattern = "_1.fastq.gz", full.names = FALSE) # Gets all names ending with _1 without full path
reverseC <- list.files("../Filtered_data", pattern = "_2.fastq.gz", full.names = FALSE) # Gets -"- ending with _2
filtFs <- file.path("../Filtered_data", "filtered", forwardC) # Creates paths for filtered forward reads
filtRs <- file.path("../Filtered_data", "filtered", reverseC) # Creates paths for filtered reverse reads

allSamples <- unique(gsub("_outFwd_1.fastq.gz|_outRev_1.fastq.gz", "", forwardC))

OutCombine <- function(dataset, samples) { # Function that summarizes rows of forward and reverse counts in the "out"-object
  counter <- 0
    for(i in samples) {
      if (counter == 0) {
        output <- data.frame()
        counter <- counter + 1
        }
      output <- rbind(output, S2 = colSums(dataset[grepl(x = rownames(dataset), pattern = i),]))
      rownames(output)[rownames(output) == "1"] <- i
      rownames(output)[rownames(output) == "S2"] <- i
    }
  names(output) <- c("reads.in", "reads.out")
  return(output)
}

out <- dada2::filterAndTrim(forward, filtFs, reverse, filtRs,
              maxN=0, truncQ=2, rm.phix=TRUE,
              compress=TRUE, multithread=TRUE) # On Windows set multithread=FALSE # maxEE=c(2,3),
saveRDS(out, file = "out.rds")
outCombined <- OutCombine(out, allSamples) # Summarize forward and reverse for every sample

forwardCounts2 <- DadaAnalysis(filtFs, filtRs)
dfForward<- as.data.frame(t(forwardCounts2))
yForward <- edgeR::DGEList(dfForward)
```

### Combine columns in "dfForward" and convert to DGEList "yAll"
```{r}
DFCombine <- function(dataset, samples) {
  counter <- 0
    for(i in samples) {
      if (counter == 0) {
        output <- data.frame(S2 = rowSums(dataset[, grepl(x = names(dataset), pattern = i)]))
        counter <- counter + 1
        names(output)[names(output) == "S2"] <- i
        next
        }
      output <- cbind(output, S2 = rowSums(dataset[, grepl(x = names(dataset), pattern = i)]))
      names(output)[names(output) == "S2"] <- i
      }
  return(output)
}

MakeDGEList <- function(dataset, samples, forwardSamples) { # Function that combines forward and reverse runs if reverse runs are present
  if(any(grepl("outRev", forwardSamples))) {
    dfAll <- DFCombine(dataset, samples)
    yAll <- edgeR::DGEList(dfAll)
  } else {
    yAll <- edgeR::DGEList(dataset)
  }
  return(yAll)
}

yAll <- MakeDGEList(dfForward, allSamples, forwardC)
saveRDS(yAll, file = "yAll.rds")
```

### Export fasta-file:
```{r}
library(Biostrings)

ExportFasta <- function(countData, fileName) {
  seqs <- row.names(countData)
  names(seqs) <- paste("Seq", 1:length(seqs), sep = "_")
  Biostrings::writeXStringSet(Biostrings::DNAStringSet(seqs, use.names = TRUE), fileName)
  sprintf("Wrote %s sequences to %s", length(seqs), fileName) 
}

ExportFasta(yAll, "y_test.fa")
```

### Parse BLAST-results:
```{r}
library(taxize)
library(gtools)

BlastParse <- function(DGEList, blastRes.out, threshold = 95) {
  sequences <- data.frame(id = paste("Seq", 1:length(rownames(DGEList)), sep = "_"), seq = row.names(DGEList))
  blastResult <- read.table(blastRes.out, sep = "\t", quote = "'", stringsAsFactors = FALSE)
  blastResult <- blastResult[blastResult$V3 > threshold,]
  names(blastResult) <- c("qseqid", "sseqid", "pident", "length", "mismatch", "gapopen", "qstart", "qend", "sstart", "send", "evalue", "bitscore", "staxids", "sscinames", "scomnames")
  blastResultUn <- blastResult[!duplicated(blastResult$qseqid),] # Retain only best hits
  
  # Slow down the query speed in a for loop to avoid time outs at NCBI
  taxonomy <- list()
  taxLength <- length(taxonomy)
  for(i in levels(factor(blastResultUn$sscinames))) {
    Sys.sleep(1)
    taxonomy[[i]] <- try(taxize::tax_name(sci = i, get = c("superkingdom", "phylum", "order", "class", "family"), db = "ncbi"))
    if(taxLength == length(taxonomy)) {
      print("Try 2")
      Sys.sleep(1)
      taxonomy[[i]] <- taxize::tax_name(sci = i, get = c("superkingdom", "phylum", "order", "class", "family"), db = "ncbi")
      if(taxLength == length(taxonomy)) {
        print("Try 3")
        Sys.sleep(1)
        taxonomy[[i]] <- taxize::tax_name(sci = i, get = c("superkingdom", "phylum", "order", "class", "family"), db = "ncbi")
      }
    }
    taxLength <- length(taxonomy)
    }
  taxonomy <- do.call("rbind", taxonomy)
  
  blastTax <- merge(blastResultUn, taxonomy, by.x = "sscinames", by.y = "query", all.x = TRUE)
  blastTax <- blastTax[, c(2,3,4,12, 1, 17:21)]
  names(blastTax) <- c("id", "besthit", "identity", "e-value", "species", "superkingdom", "phylum", "order", "class", "family")
  seqTax <- merge(sequences, blastTax, by.x = "id", by.y = "id", all.x = TRUE)
  seqTax$id <- as.character(seqTax$id)
  seqTax <- seqTax[mixedorder(seqTax$id),]
  seqTax$superkingdom <- factor(seqTax$superkingdom)
  seqTax$phylum <- factor(seqTax$phylum)
  seqTax$order <- factor(seqTax$order)
  seqTax$class <- factor(seqTax$class)
  seqTax$family <- factor(seqTax$family)
  return(seqTax)
}

blastResY <- BlastParse(yAll, "y_test.out")
saveRDS(blastResY, file = "blastResY.rds")
```

### Create Summaryfiles from the BLAST-results:
```{r}
grepGroup <- grepl("Arthropoda", blastResY$phylum) # Set which group to put in "groupSummary"

genesCounts <- cbind(blastResY, yAll$counts)
summaryAll <- aggregate(genesCounts[,12:dim(genesCounts)[2]], by = list(genesCounts$species), FUN = sum)
summaryAll <- summaryAll[order(rowSums(summaryAll[,-1]), decreasing = TRUE),]
genesGroup <- blastResY[grepGroup,]
yGroup <- yAll[row.names(yAll$counts) %in% genesGroup$seq,]
genesCountsGroup <- cbind(genesGroup, yGroup$counts)
groupSummary <- aggregate(genesCountsGroup[,12:dim(genesCounts)[2]], by = list(genesCountsGroup$species), FUN=sum)
groupSummary <- groupSummary[order(rowSums(groupSummary[,-1]), decreasing = TRUE),]

saveRDS(yAll, file = "yAll.rds")
saveRDS(genesCounts, file = "genesCounts.rds")
saveRDS(groupSummary, file = "groupSummary.rds")
saveRDS(summaryAll, file = "summaryAll.rds")
```

### Functions:
```{r}
RenameSpecies <- function(oldName, newName, dataFrame) {
  dataFrame[grepl(oldName, dataFrame[,1]),1] <- newName
  return(dataFrame)
}

CombineSpecies <- function(species1, species2, dataFrame) {
  speciesSum <- dataFrame[grepl(species1, dataFrame[,4]),5:ncol(dataFrame)] + dataFrame[grepl(species2, dataFrame[,4]),5:ncol(dataFrame)]
  speciesSum <- cbind(dataFrame[grepl(species1, dataFrame[,4]),1:4], speciesSum)
  names(speciesSum)[1:4] <- names(dataFrame[1:4])
  dataFrame <- dataFrame[!grepl(species1, dataFrame[,4]),]
  dataFrame <- dataFrame[!grepl(species2, dataFrame[,4]),]
  dataFrame <- rbind(dataFrame, speciesSum)
  return(dataFrame)
}

SpeciesPercent <- function(dataFrame) {
  percentVector <- c()
  for(row in 1:nrow(dataFrame)) {
    ratio <- sum(dataFrame[row,2:ncol(dataFrame)]) / sum(dataFrame[,2:ncol(dataFrame)])
    percent <- ratio * 100
    percentVector <- c(percentVector, round(percent, 5))
  }
  return(percentVector)
}

QSeqGetter <- function(species_name) {
  species_seq <- paste0(">", blastResY[grepl(species_name, blastResY$species),1], "\n", blastResY[grepl(species_name, blastResY$species),2])
  writeLines(species_seq, "current_sequence.txt")
  file.edit("current_sequence.txt")
}

RemoveLowFreqSeqs <- function(dataset, threshold, subValue = 0) {
  output <- data.frame(dataset[, 1:4])
  for(column in 5:ncol(dataset)){
    ratio <- dataset[, column] / sum(dataset[, column]) * 100
    dataset[ratio < threshold, column] <- subValue
    output <- cbind(output, dataset[, column])
  }
  colnames(output) <- colnames(dataset)
  return(output)
}
```

### Curate the groupSummary-object:
```{r}
library(MetaBAnalysis)
{
  groupNotes <- groupSummary
  # groupNotes <- RenameSpecies("old latin name", "new latin name", groupNotes)
  groupNotes <- groupNotes[order(rowSums(groupNotes[,-1]), decreasing = TRUE),]
  groupNotes <- cbind(Percent_all = SpeciesPercent(groupNotes), groupNotes)
  groupNotes <- cbind(Swedish = MetaBAnalysis::TranslateTaxa(groupNotes$Group.1), groupNotes)
  # groupNotes[grepl("latin name", groupNotes$Group.1),1] <- "new common name"
  groupNotes <- cbind(Notes = c(
                                # edit here to correspond to the number of species in the groupSummary-object
				"", "", "", "", "",
                                "", "", "", "", "",
                                "", "", "", "", "",
                                "", "", "", "", ""
                                ), groupNotes) # add notes for different species
  # groupNotes <- CombineSpecies("species to be kept", "species to be absorbed", groupNotes)
  groupNotes <- groupNotes[order(rowSums(groupNotes[,-c(1, 2, 3, 4)]), decreasing = TRUE),]
  # groupNotes <- groupNotes[!grepl("drop", groupNotes$Notes),] # Drop all species that has the word "drop" in their notes
  # groupNotes <- RemoveLowFreqSeqs(groupNotes, 0.5) # remove species within samples below a threshold
  # groupNotes <- groupNotes[groupNotes$Percent_all >= 0.5,] # remove species from all samples below a threshold
  groupClean <- groupNotes[,-c(1, 3)]
  columnNames <- colnames(groupClean)
  columnNames[grepl("Group.1", columnNames)] <- "Latin"
  # columnNames[grepl("old sample name", columnNames)] <- "new sample name"
  names(groupClean) <- columnNames
}

saveRDS(groupClean, "groupClean.rds") # save the clean object to be used in the report
```
